mm2src/coins/eth/eth_swap_v2/eth_taker_swap_v2.rs

```diff
@@ -457,34 +458,88 @@ impl EthCoin {
-                    &TAKER_SWAP_V2,
-                    EthPaymentType::TakerPayments,
-                    TAKER_PAYMENT_STATE_INDEX,
-                )
-                .await?;
-            if taker_status == U256::from(TakerPaymentStateV2::MakerSpent as u8) {
-                return Ok(taker_payment.clone());
-            }
             let now = now_sec();
             if now > wait_until {
-                return MmError::err(WaitForPaymentSpendError::Timeout { wait_until, now });
+                return MmError::err(FindPaymentSpendError::Timeout { wait_until, now });
+            }
+
+            // Skip retrieving events if tx_hash is already found
+            if tx_hash.is_none() {
+                // get all logged TakerPaymentSpent events from `from_block` till current block
+                let events = match self
+                    .events_from_block(taker_swap_v2_contract, "TakerPaymentSpent", from_block, &TAKER_SWAP_V2)
+                    .await
+                {
+                    Ok(events) => events,
+                    Err(e) => {
+                        error!(
+                            "Error getting TakerPaymentSpent events from {} block: {}",
+                            from_block, e
+                        );
+                        Timer::sleep(check_every).await;
+                        continue;
+                    },
+                };
+
+                if events.is_empty() {
+                    info!("No events found yet from block {}", from_block);
+                    Timer::sleep(check_every).await;
+                    continue;
+                }
+
+                // this is how spent event looks like in EtomicSwapTakerV2: event TakerPaymentSpent(bytes32 id, bytes32 secret)
+                let found_event = events.into_iter().find(|event| &event.data.0[..32] == id.as_slice());
+
+                if let Some(event) = found_event {
+                    if let Some(hash) = event.transaction_hash {
+                        // Store tx_hash to skip fetching events in the next iteration if "eth_getTransactionByHash" is unsuccessful
+                        tx_hash = Some(hash);
+                    }
+                }
             }
-            Timer::sleep(10.).await;
+
+            // Proceed to check transaction if we have a tx_hash
+            if let Some(tx_hash) = tx_hash {
+                match self.transaction(TransactionId::Hash(tx_hash)).await {
+                    Ok(Some(t)) => {
+                        let transaction = signed_tx_from_web3_tx(t).map_err(FindPaymentSpendError::Internal)?;
+                        return Ok(transaction);
+                    },
+                    Ok(None) => info!("spendTakerPayment transaction {} not found yet", tx_hash),
+                    Err(e) => error!("Get spendTakerPayment transaction {} error: {}", tx_hash, e),
+                };
+                Timer::sleep(check_every).await;
+                continue;
```

Comment:
these 2 lines just could be deleted btw

Reply:
I suggest leaving them. If someone in the future adds logic after the `if let Some(tx_hash) = tx_hash` block, they might forget to add the `continue` again, or they might decide that not having the `continue` and sleep time is what we actually want. It's better to avoid confusion.![image](https://github.com/user-attachments/assets/3d156d11-bf2e-4f34-ba06-9f09a65ed946)

Reply:
> I suggest leaving them. If someone in the future adds logic after the if let Some(tx_hash) = tx_hash block, they might forget to add the continue again, or they might decide that not having the continue and sleep time is what we actually want. It's better to avoid confusion.

no, whoever's adding the code should understand the how the current flow executes and how the new code will execute along.
leaving this would be confusing tbh, i would see this and question if it really does something important and think i misunderstand something (i would prolly double check what `continue` does in rust, that's the level of confusion i mean).

Reply:
> whoever's adding the code should understand the how the current flow executes and how the new code will execute along.

should, but the human factor must be taken into account 

>  if it really does something important 

It does. If the transaction wasn't found by tx_hash, the next loop cycle should be started. Its better to use `continue;` explicitly in such places.

Reply:
we can remove time sleep from the end of loop https://github.com/KomodoPlatform/komodo-defi-framework/pull/2261/commits/2056f9bfd9c7890d09657d9f92052f0613eb80a1